src/extra/components/xgm.py

```diff
@@ -0,0 +1,658 @@
+        """Return a 1D [DataArray][xarray.DataArray] of the nominal wavelength
+        of each train.
+        """
+        if self._wavelength_by_train is None:
+            self._wavelength_by_train = self._control_source["pulseEnergy.wavelengthUsed"].xarray()
+            self._wavelength_by_train.attrs["units"] = "nm"
+
+        return self._wavelength_by_train
+
+    def photon_energy(self, with_units=True):
+        """The nominal photon energy in keV.
+
+        This calls
+        [KeyData.as_single_value()][extra_data.KeyData.as_single_value]
+        internally, which means it will throw an exception if the wavelength is
+        not constant.
+
+        Args:
+            with_units (bool): Whether to return a [pint.Quantity][] or a [float][].
+        """
+        energy = self.wavelength().to("keV", "xfel")
+        return energy if with_units else energy.magnitude
+
+    def photon_energy_by_train(self):
+        """The nominal photon energy in keV of each train."""
+        if self._photon_energy_by_train is None:
+            import xarray as xr
+
+            wavelengths = self.wavelength_by_train()
+            energies = (wavelengths * ureg.nm).data.to("keV", "xfel").magnitude
+            energies = xr.DataArray(energies,
+                                    dims=("trainId",),
+                                    coords=dict(trainId=wavelengths.trainId),
+                                    attrs=dict(units="keV"))
+            self._photon_energy_by_train = energies
+
+        return self._photon_energy_by_train
+
+    def doocs_server(self) -> str:
+        """The DOOCS server that the Karabo XGM device is connected to.
+
+        This information was not stored in older versions of the XGM, so the
+        method may return `None` if the DOOCS server was not found.
+
+        Raises:
+            extra_data.MultiRunError: A `MultiRunError` will be thrown whenever
+                the [DataCollection][extra_data.DataCollection] contains data
+                from more than one run. This XGM property is stored in the `RUN`
+                section of EXDF files, and when a run has been
+                [unioned][extra_data.DataCollection.union] with another it is
+                not possible to retrieve `RUN` values.
+        """
+        if self._doocs_server is None:
+            self._doocs_server = self._control_source.run_values().get("location.value")
+
+        return self._doocs_server
+
+    def _check_sase_arg(self, sase):
+        """Helper function to return the correct PropertyGroup."""
+        if len(self._extra_sases) == 0 and sase is not None:
+            raise RuntimeError("This XGM doesn't have data from any other SASEs")
+        if sase is not None and sase not in [0, 1, 3]:
+            raise RuntimeError(f"Invalid SASE '{sase}', it must be either 0, 1, or 3")
+
+        return PropertyGroup(sase) if sase is not None else self._default_pg
+
+    def intensity(self, sase=None):
```

Comment:
Sorry if I'm late with this, but looking over it again I just realized that this is physically wrong. [Intensity](https://en.wikipedia.org/wiki/Intensity) is defined as power per area, while here we're looking at [pulse] energy.

I realize this term is used colloquially sometimes (and even some properties are wrong, e.g. `photonFlux`, which is also defined differently).

Strong feelings against `pulse_energy`?